lib/ecmascript.ts

```diff
@@ -6754,112 +6374,58 @@ export function RoundDuration(
       if (oneWeekDays === 0) throw new RangeError('custom calendar reported that a week is 0 days long');
       // dayLengthNs is never undefined if unit is `day` or larger.
       assertExists(dayLengthNs);
-      const divisor = JSBI.multiply(JSBI.BigInt(oneWeekDays), dayLengthNs);
-      nanoseconds = JSBI.add(
-        JSBI.add(JSBI.multiply(divisor, JSBI.BigInt(weeks)), JSBI.multiply(JSBI.BigInt(days), dayLengthNs)),
-        nanoseconds
-      );
-      const rounded = RoundNumberToIncrement(nanoseconds, JSBI.multiply(divisor, JSBI.BigInt(increment)), roundingMode);
-      total = BigIntDivideToNumber(nanoseconds, divisor);
-      weeks = JSBI.toNumber(JSBI.divide(rounded, divisor));
-      nanoseconds = ZERO;
+      total = weeks + (days + norm.fdiv(dayLengthNs)) / oneWeekDays;
+      weeks = RoundJSNumberToIncrement(total, increment, roundingMode);
       days = 0;
+      norm = TimeDuration.ZERO;
       break;
     }
     case 'day': {
       // dayLengthNs is never undefined if unit is `day` or larger.
       assertExists(dayLengthNs);
-      const divisor = dayLengthNs;
-      nanoseconds = JSBI.add(JSBI.multiply(divisor, JSBI.BigInt(days)), nanoseconds);
-      const rounded = RoundNumberToIncrement(nanoseconds, JSBI.multiply(divisor, JSBI.BigInt(increment)), roundingMode);
-      total = BigIntDivideToNumber(nanoseconds, divisor);
-      days = JSBI.toNumber(JSBI.divide(rounded, divisor));
-      nanoseconds = ZERO;
+      total = days + norm.fdiv(dayLengthNs);
+      days = RoundJSNumberToIncrement(total, increment, roundingMode);
+      norm = TimeDuration.ZERO;
       break;
     }
     case 'hour': {
       const divisor = 3600e9;
-      let allNanoseconds = JSBI.multiply(JSBI.BigInt(hours), JSBI.BigInt(3600e9));
-      allNanoseconds = JSBI.add(allNanoseconds, JSBI.multiply(JSBI.BigInt(minutes), JSBI.BigInt(60e9)));
-      allNanoseconds = JSBI.add(allNanoseconds, JSBI.multiply(JSBI.BigInt(seconds), BILLION));
-      allNanoseconds = JSBI.add(allNanoseconds, JSBI.multiply(JSBI.BigInt(milliseconds), MILLION));
-      allNanoseconds = JSBI.add(allNanoseconds, JSBI.multiply(JSBI.BigInt(microseconds), THOUSAND));
-      allNanoseconds = JSBI.add(allNanoseconds, nanoseconds);
-      total = BigIntDivideToNumber(allNanoseconds, JSBI.BigInt(divisor));
-      const rounded = RoundNumberToIncrement(allNanoseconds, JSBI.BigInt(divisor * increment), roundingMode);
-      hours = JSBI.toNumber(JSBI.divide(rounded, JSBI.BigInt(divisor)));
-      nanoseconds = ZERO;
-      minutes = seconds = milliseconds = microseconds = 0;
+      total = norm.fdiv(divisor);
+      norm = norm.round(JSBI.BigInt(divisor * increment), roundingMode);
```

Comment:
Should `round` take a Number instead of a JSBI as its first argument?

Reply:
I agree. I thought it'd be better to accept a JSBI since we use JSBI inside the `round` method, but on reflection accepting a Number and converting will encapsulate the bigint usage inside TimeDuration, which is good.